File: 2020/knowit/5/5.js
"use strict";

const fs = require("fs");

const path = fs.readFileSync("rute.txt", "utf-8");

const findVertices = path => {
	let x = 0;
	let y = 0;

	const vertices = [];

	for (let i = 0; i < path.length; i++) {
		// XXX Exploits the fact that path[-1] is undefined to make the first vertex. Also ignores the last one
		if (path[i-1] !== path[i]) {
			vertices.push([x, y]);
		}

		if (path[i] === "H") {
			x++;
		} else if (path[i] === "V") {
			x--;
		} else if (path[i] === "O") {
			y--;
		} else if (path[i] === "N") {
			y++;
		}
	}

	return vertices;
};

const calculateArea = vertices => {
	const X = 0;
	const Y = 1;
	let area = 0;
	let j = vertices.length-1;
	for (let i = 0; i < vertices.length; i++) {
		area += (vertices[j][X] + vertices[i][X]) * (vertices[j][Y] - vertices[i][Y]);
		j = i;
	}

	return Math.abs(area) / 2;
};

console.log(calculateArea(findVertices(path)));

File: 2020/knowit/5/5.hs
import System.IO (readFile)

type Point = (Int, Int)

findVertices :: String -> [Point]
findVertices s = (0,0) : findVertices' (0,0) s

findVertices' :: Point -> String -> [Point]
findVertices' p [] = [p]
findVertices' (x,y) as@(a:_)
    | a == 'H' = something (x+1,y) as
    | a == 'V' = something (x-1,y) as
    | a == 'O' = something (x,y-1) as
    | a == 'N' = something (x,y+1) as
    where something p (a1:[]) = []
          something p (a1:a2:as)
              | a1 /= a2  = p : findVertices' p (a2:as)
              | otherwise = findVertices' p (a2:as)

calculateArea :: [Point] -> Int
calculateArea vs =
    abs
    . (`div` 2)
    . sum
    $ zipWith det (last vs : init vs) vs
        where det (x1,y1) (x2,y2) = (x2+x1) * (y2-y1)

main =
    print
    . calculateArea
    . findVertices
    =<< readFile "rute.txt"
    -- $ "HHOOVVNN"
    -- $ "HHHHHHOOOOVVNNNVVOVVNN"
